File: pet-allergy-scanner/SniffTest/Features/Nutrition/Services/FoodComparisonService.swift
//
//  FoodComparisonService.swift
//  SniffTest
//
//  Created by Steven Matos on 9/26/25.
//

import Foundation
import Combine

/**
 * Food Comparison Service
 * 
 * Handles all food comparison operations including:
 * - Side-by-side food comparison
 * - Nutritional value analysis
 * - Cost per nutritional value analysis
 * - Recommendation generation
 * - Comparison history management
 * 
 * Follows SOLID principles with single responsibility for food comparison
 * Implements DRY by reusing common analysis methods
 * Follows KISS by keeping the API simple and focused
 */
@MainActor
class FoodComparisonService: ObservableObject {
    static let shared = FoodComparisonService()
    
    @Published var recentComparisons: [SavedComparison] = []
    @Published var currentComparison: FoodComparisonResults?
    @Published var isLoading = false
    @Published var error: Error?
    
    private let apiService: APIService
    private var cancellables = Set<AnyCancellable>()
    
    private init() {
        self.apiService = APIService.shared
        loadRecentComparisons()
    }
    
    // MARK: - Public API
    
    /**
     * Compare multiple foods
     * - Parameter foodIds: Array of food IDs to compare
     * - Parameter comparisonName: Name for the comparison
     * - Parameter pet: Optional pet to check for allergies/sensitivities
     * - Returns: Comparison results
     */
    func compareFoods(foodIds: [String], comparisonName: String, pet: Pet? = nil) async throws -> FoodComparisonResults {
        guard foodIds.count >= 2 else {
            throw ComparisonError.insufficientFoods
        }
        
        guard foodIds.count <= 3 else {
            throw ComparisonError.tooManyFoods
        }
        
        isLoading = true
        error = nil
        
        do {
            // Load food details
            var foods = try await loadFoodDetails(foodIds: foodIds)
            
            // Check for pet allergies if pet is provided
            if let pet = pet {
                foods = checkPetAllergies(foods: foods, pet: pet)
            }
            
            // Generate comparison metrics
            let metrics = generateComparisonMetrics(foods: foods)
            
            // Generate recommendations
            let recommendations = generateRecommendations(metrics: metrics, foods: foods, pet: pet)
            
            // Determine best options
            let bestOptions = determineBestOptions(metrics: metrics, foods: foods)
            
            // Create comparison results
            let results = FoodComparisonResults(
                id: UUID().uuidString,
                comparisonName: comparisonName,
                foods: foods,
                bestOverall: bestOptions.overall,
                bestValue: bestOptions.value,
                bestNutrition: bestOptions.nutrition,
                costPerCalorie: metrics.costPerCalorie,
                nutritionalDensity: metrics.nutritionalDensity,
                compatibilityScores: metrics.compatibilityScores,
                recommendations: recommendations,
                petAllergiesChecked: pet != nil,
                petName: pet?.name
            )
            
            // Save to recent comparisons
            await saveComparison(results)
            
            currentComparison = results
            isLoading = false
            
            return results
            
        } catch {
            self.error = error
            isLoading = false
            throw error
        }
    }
    
    /**
     * Load a saved comparison from backend
     * - Parameter comparisonId: ID of the comparison to load
     * - Returns: Comparison results
     */
    func loadComparison(comparisonId: String) async throws -> FoodComparisonResults {
        isLoading = true
        error = nil
        
        do {
            // Fetch comparison from backend
            struct ComparisonResponse: Codable {
                let id: String
                let comparisonName: String
                let foodIds: [String]
                let comparisonData: ComparisonData
                
                struct ComparisonData: Codable {
                    let foods: [FoodData]
                    
                    struct FoodData: Codable {
                        let id: String
                        let name: String
                        let brand: String?
                        let caloriesPer100g: Double
                        let proteinPercentage: Double
                        let fatPercentage: Double
                        let fiberPercentage: Double
                        let moisturePercentage: Double
                        let ingredients: [String]
                        let allergens: [String]
                        
                        enum CodingKeys: String, CodingKey {
                            case id, name, brand, ingredients, allergens
                            case caloriesPer100g = "calories_per_100g"
                            case proteinPercentage = "protein_percentage"
                            case fatPercentage = "fat_percentage"
                            case fiberPercentage = "fiber_percentage"
                            case moisturePercentage = "moisture_percentage"
                        }
                    }
                }
                
                enum CodingKeys: String, CodingKey {
                    case id
                    case comparisonName = "comparison_name"
                    case foodIds = "food_ids"
                    case comparisonData = "comparison_data"
                }
            }
            
            let response = try await apiService.get(
                endpoint: "/advanced-nutrition/comparisons/\(comparisonId)",
                responseType: ComparisonResponse.self
            )
            
            // Convert to FoodAnalysis objects
            let foods = response.comparisonData.foods.map { food in
                FoodAnalysis(
                    id: food.id,
                    petId: "",
                    foodName: food.name,
                    brand: food.brand,
                    caloriesPer100g: food.caloriesPer100g,
                    proteinPercentage: food.proteinPercentage,
                    fatPercentage: food.fatPercentage,
                    fiberPercentage: food.fiberPercentage,
                    moisturePercentage: food.moisturePercentage,
                    ingredients: food.ingredients,
                    allergens: food.allergens,
                    analyzedAt: Date()
                )
            }
            
            // Generate metrics and recommendations
            let metrics = generateComparisonMetrics(foods: foods)
            let recommendations = generateRecommendations(metrics: metrics, foods: foods, pet: nil)
            let bestOptions = determineBestOptions(metrics: metrics, foods: foods)
            
            let results = FoodComparisonResults(
                id: response.id,
                comparisonName: response.comparisonName,
                foods: foods,
                bestOverall: bestOptions.overall,
                bestValue: bestOptions.value,
                bestNutrition: bestOptions.nutrition,
                costPerCalorie: metrics.costPerCalorie,
                nutritionalDensity: metrics.nutritionalDensity,
                compatibilityScores: metrics.compatibilityScores,
                recommendations: recommendations,
                petAllergiesChecked: false,
                petName: nil
            )
            
            currentComparison = results
            isLoading = false
            
            return results
            
        } catch {
            self.error = error
            isLoading = false
            throw error
        }
    }
    
    /**
     * Delete a saved comparison
     * - Parameter comparisonId: ID of the comparison to delete
     */
    func deleteComparison(comparisonId: String) async throws {
        do {
            // Delete from backend
            try await apiService.delete(
                endpoint: "/advanced-nutrition/comparisons/\(comparisonId)"
            )
            
            // Remove from local storage
            recentComparisons.removeAll { $0.id == comparisonId }
            
        } catch {
            throw error
        }
    }
    
    /**
     * Get comparison history
     * - Returns: Array of saved comparisons
     */
    func getComparisonHistory() -> [SavedComparison] {
        return recentComparisons
    }
    
    // MARK: - Private Methods
    
    /**
     * Load food details from backend
     * - Parameter foodIds: Array of food IDs to load
     * - Returns: Array of food analysis objects
     */
    private func loadFoodDetails(foodIds: [String]) async throws -> [FoodAnalysis] {
        var foodAnalyses: [FoodAnalysis] = []
        
        // Load each food item
        for foodId in foodIds {
            do {
                let foodItem = try await apiService.get(
                    endpoint: "/foods/\(foodId)",
                    responseType: FoodItem.self
                )
                
                // Convert FoodItem to FoodAnalysis
                let analysis = FoodAnalysis(
                    id: foodItem.id,
                    petId: "", // Not pet-specific for comparison
                    foodName: foodItem.name,
                    brand: foodItem.brand,
                    caloriesPer100g: foodItem.nutritionalInfo?.caloriesPer100g ?? 0,
                    proteinPercentage: foodItem.nutritionalInfo?.proteinPercentage ?? 0,
                    fatPercentage: foodItem.nutritionalInfo?.fatPercentage ?? 0,
                    fiberPercentage: foodItem.nutritionalInfo?.fiberPercentage ?? 0,
                    moisturePercentage: foodItem.nutritionalInfo?.moisturePercentage ?? 0,
                    ingredients: foodItem.nutritionalInfo?.ingredients ?? [],
                    allergens: foodItem.nutritionalInfo?.allergens ?? [],
                    analyzedAt: foodItem.updatedAt
                )
                
                foodAnalyses.append(analysis)
            } catch {
                throw ComparisonError.foodNotFound(foodId)
            }
        }
        
        return foodAnalyses
    }
    
    private func generateComparisonMetrics(foods: [FoodAnalysis]) -> ComparisonMetrics {
        var costPerCalorie: [String: Double] = [:]
        var nutritionalDensity: [String: Double] = [:]
        var compatibilityScores: [String: Double] = [:]
        
        for (index, food) in foods.enumerated() {
            // Mock cost per calorie calculation
            costPerCalorie[food.id] = 0.01 + (Double(index) * 0.005)
            
            // Mock nutritional density calculation
            let essentialNutrients = food.proteinPercentage + food.fatPercentage
            nutritionalDensity[food.id] = essentialNutrients > 0 ? food.caloriesPer100g / essentialNutrients : 0
            
            // Mock compatibility score
            compatibilityScores[food.id] = 70.0 + (Double(index) * 5.0) + Double.random(in: -10...10)
        }
        
        return ComparisonMetrics(
            costPerCalorie: costPerCalorie,
            nutritionalDensity: nutritionalDensity,
            compatibilityScores: compatibilityScores
        )
    }
    
    /**
     * Check foods against pet's known allergies and sensitivities
     * Updates each food with allergy warnings
     */
    private func checkPetAllergies(foods: [FoodAnalysis], pet: Pet) -> [FoodAnalysis] {
        return foods.map { food in
            var updatedFood = food
            var matchedAllergens: [String] = []
            
            // Check each ingredient against pet's known sensitivities
            let petSensitivities = Set(pet.knownSensitivities.map { $0.lowercased().trimmingCharacters(in: .whitespaces) })
            
            for ingredient in food.ingredients {
                let normalizedIngredient = ingredient.lowercased().trimmingCharacters(in: .whitespaces)
                
                // Check for exact matches
                if petSensitivities.contains(normalizedIngredient) {
                    matchedAllergens.append(ingredient)
                    continue
                }
                
                // Check for partial matches (e.g., "chicken meal" matches "chicken")
                for sensitivity in petSensitivities {
                    if normalizedIngredient.contains(sensitivity) || sensitivity.contains(normalizedIngredient) {
                        matchedAllergens.append(ingredient)
                        break
                    }
                }
            }
            
            updatedFood.petAllergyWarnings = matchedAllergens
            updatedFood.hasPetAllergyWarning = !matchedAllergens.isEmpty
            
            return updatedFood
        }
    }
    
    private func generateRecommendations(metrics: ComparisonMetrics, foods: [FoodAnalysis], pet: Pet?) -> [String] {
        var recommendations: [String] = []
        
        // Pet allergy warnings (MOST IMPORTANT - show first)
        if let pet = pet {
            let foodsWithAllergens = foods.filter { $0.hasPetAllergyWarning }
            if !foodsWithAllergens.isEmpty {
                recommendations.append("⚠️ WARNING: \(foodsWithAllergens.count) food(s) contain ingredients that \(pet.name) is sensitive to")
                
                // List specific foods with allergens
                for food in foodsWithAllergens {
                    if let allergens = food.petAllergyWarnings, !allergens.isEmpty {
                        recommendations.append("• \(food.foodName): Contains \(allergens.joined(separator: ", "))")
                    }
                }
            } else {
                recommendations.append("✓ All foods are safe for \(pet.name)'s known sensitivities")
            }
        }
        
        // Protein analysis
        let proteinValues = foods.map { $0.proteinPercentage }
        if let maxProtein = proteinValues.max(), let minProtein = proteinValues.min() {
            if maxProtein - minProtein > 10 {
                recommendations.append("Protein content varies significantly (\(String(format: "%.1f", minProtein))% - \(String(format: "%.1f", maxProtein))%)")
            }
        }
        
        // Calorie analysis
        let calorieValues = foods.map { $0.caloriesPer100g }
        if let maxCalories = calorieValues.max(), let minCalories = calorieValues.min() {
            if maxCalories - minCalories > 100 {
                recommendations.append("Calorie content varies significantly (\(String(format: "%.0f", minCalories)) - \(String(format: "%.0f", maxCalories)) kcal/100g)")
            }
        }
        
        return recommendations
    }
    
    private func determineBestOptions(metrics: ComparisonMetrics, foods: [FoodAnalysis]) -> BestOptions {
        // CRITICAL: Filter out foods with pet sensitivities first!
        // Never recommend a food that clashes with pet's known sensitivities
        let safeFoods = foods.filter { !$0.hasPetAllergyWarning }
        
        // If all foods have sensitivities, return warning message
        guard !safeFoods.isEmpty else {
            return BestOptions(
                overall: "⚠️ All foods contain pet sensitivities",
                value: "⚠️ All foods contain pet sensitivities",
                nutrition: "⚠️ All foods contain pet sensitivities"
            )
        }
        
        // Best overall (highest composite score) - only from safe foods
        let overallScores = safeFoods.map { food in
            let calorieScore = min(100, max(0, 100 - abs(food.caloriesPer100g - 350) / 3.5))
            let proteinScore = min(100, food.proteinPercentage * 2)
            let fatScore = min(100, food.fatPercentage * 4)
            let fiberScore = min(100, food.fiberPercentage * 10)
            let costScore = min(100, max(0, 100 - (metrics.costPerCalorie[food.id] ?? 0) * 1000))
            let compatibilityScore = metrics.compatibilityScores[food.id] ?? 0
            
            return (food, calorieScore * 0.25 + proteinScore * 0.25 + fatScore * 0.20 + fiberScore * 0.15 + costScore * 0.15 + compatibilityScore * 0.15)
        }
        
        let bestOverall = overallScores.max { $0.1 < $1.1 }?.0.foodName ?? "Unknown"
        
        // Best value (lowest cost per calorie) - only from safe foods
        let bestValue = safeFoods.min { 
            (metrics.costPerCalorie[$0.id] ?? Double.infinity) < (metrics.costPerCalorie[$1.id] ?? Double.infinity) 
        }?.foodName ?? "Unknown"
        
        // Best nutrition (highest protein) - only from safe foods
        let bestNutrition = safeFoods.max { 
            $0.proteinPercentage < $1.proteinPercentage
        }?.foodName ?? "Unknown"
        
        return BestOptions(
            overall: bestOverall,
            value: bestValue,
            nutrition: bestNutrition
        )
    }
    
    /**
     * Save comparison to backend
     * - Parameter results: Comparison results to save
     */
    private func saveComparison(_ results: FoodComparisonResults) async {
        do {
            // Create comparison request
            struct ComparisonRequest: Codable {
                let comparisonName: String
                let foodIds: [String]
                
                enum CodingKeys: String, CodingKey {
                    case comparisonName = "comparison_name"
                    case foodIds = "food_ids"
                }
            }
            
            struct ComparisonResponse: Codable {
                let id: String
                let comparisonName: String
                let createdAt: String
                
                enum CodingKeys: String, CodingKey {
                    case id
                    case comparisonName = "comparison_name"
                    case createdAt = "created_at"
                }
            }
            
            let request = ComparisonRequest(
                comparisonName: results.comparisonName,
                foodIds: results.foods.map { $0.id }
            )
            
            let response = try await apiService.post(
                endpoint: "/advanced-nutrition/comparisons",
                body: request,
                responseType: ComparisonResponse.self
            )
            
            // Add to recent comparisons
            let savedComparison = SavedComparison(
                id: response.id,
                name: response.comparisonName,
                foodCount: results.foods.count,
                createdAt: Date()
            )
            
            recentComparisons.insert(savedComparison, at: 0)
            
            // Keep only last 20 comparisons
            if recentComparisons.count > 20 {
                recentComparisons = Array(recentComparisons.prefix(20))
            }
            
        } catch {
            print("Failed to save comparison to backend: \(error)")
            // Still add to local cache even if backend save fails
            let savedComparison = SavedComparison(
                id: results.id,
                name: results.comparisonName,
                foodCount: results.foods.count,
                createdAt: Date()
            )
            
            recentComparisons.insert(savedComparison, at: 0)
            if recentComparisons.count > 20 {
                recentComparisons = Array(recentComparisons.prefix(20))
            }
        }
    }
    
    /**
     * Load recent comparisons from backend
     */
    private func loadRecentComparisons() {
        Task {
            do {
                struct ComparisonListItem: Codable {
                    let id: String
                    let comparisonName: String
                    let foodIds: [String]
                    let createdAt: String
                    
                    enum CodingKeys: String, CodingKey {
                        case id
                        case comparisonName = "comparison_name"
                        case foodIds = "food_ids"
                        case createdAt = "created_at"
                    }
                }
                
                let response = try await apiService.get(
                    endpoint: "/advanced-nutrition/comparisons?limit=20",
                    responseType: [ComparisonListItem].self
                )
                
                let comparisons = response.compactMap { item -> SavedComparison? in
                    // Parse ISO8601 date
                    let formatter = ISO8601DateFormatter()
                    guard let date = formatter.date(from: item.createdAt) else {
                        return nil
                    }
                    
                    return SavedComparison(
                        id: item.id,
                        name: item.comparisonName,
                        foodCount: item.foodIds.count,
                        createdAt: date
                    )
                }
                
                await MainActor.run {
                    recentComparisons = comparisons
                }
                
            } catch {
                print("Failed to load recent comparisons: \(error)")
                // Use empty array if load fails
                await MainActor.run {
                    recentComparisons = []
                }
            }
        }
    }
}

// MARK: - Data Models

struct FoodComparisonResults {
    let id: String
    let comparisonName: String
    let foods: [FoodAnalysis]
    let bestOverall: String
    let bestValue: String
    let bestNutrition: String
    let costPerCalorie: [String: Double]
    let nutritionalDensity: [String: Double]
    let compatibilityScores: [String: Double]
    let recommendations: [String]
    let petAllergiesChecked: Bool
    let petName: String?
}

struct ComparisonMetrics {
    let costPerCalorie: [String: Double]
    let nutritionalDensity: [String: Double]
    let compatibilityScores: [String: Double]
}

struct BestOptions {
    let overall: String
    let value: String
    let nutrition: String
}

struct FoodAnalysis: Identifiable {
    let id: String
    let petId: String
    let foodName: String
    let brand: String?
    let caloriesPer100g: Double
    let proteinPercentage: Double
    let fatPercentage: Double
    let fiberPercentage: Double
    let moisturePercentage: Double
    let ingredients: [String]
    let allergens: [String]
    let analyzedAt: Date
    var hasPetAllergyWarning: Bool = false
    var petAllergyWarnings: [String]? = nil
}

// MARK: - Errors

enum ComparisonError: Error, LocalizedError {
    case insufficientFoods
    case tooManyFoods
    case foodNotFound(String)
    case notImplemented
    
    var errorDescription: String? {
        switch self {
        case .insufficientFoods:
            return "At least 2 foods are required for comparison"
        case .tooManyFoods:
            return "Maximum 3 foods allowed for comparison"
        case .foodNotFound(let foodId):
            return "Food with ID \(foodId) not found"
        case .notImplemented:
            return "Feature not yet implemented"
        }
    }
}

// MARK: - API Service Extensions
// Removed placeholder methods - now using real API endpoints

File: server/app/services/nutrition/food_comparison_service.py
"""
Food Comparison Service
Handles side-by-side food comparison, nutritional analysis, and recommendations
"""

from typing import List, Optional, Dict, Any, Tuple
from datetime import datetime
from app.shared.services.datetime_service import DateTimeService
from decimal import Decimal
import statistics
from supabase import Client
from app.shared.services.database_operation_service import DatabaseOperationService
from app.models.nutrition.advanced_nutrition import (
    FoodComparisonCreate, FoodComparisonResponse,
    FoodComparisonMetrics, FoodComparisonDashboard
)
from app.models.nutrition.nutrition import FoodAnalysisResponse


class FoodComparisonService:
    """
    Service for comparing pet foods and generating recommendations
    
    Follows SOLID principles with single responsibility for food comparison
    Implements DRY by reusing common calculation methods
    Follows KISS by keeping comparison logic focused and simple
    """
    
    def __init__(self, supabase: Client):
        """
        Initialize food comparison service
        
        Args:
            supabase: Authenticated Supabase client (for RLS compliance)
        """
        self.supabase = supabase
    
    async def create_comparison(
        self, 
        comparison: FoodComparisonCreate,
        user_id: str
    ) -> FoodComparisonResponse:
        """
        Create a new food comparison
        
        Args:
            comparison: Comparison data
            user_id: User ID for authorization
            
        Returns:
            Created comparison
        """
        # Validate food IDs exist and user has access
        await self._validate_food_access(comparison.food_ids, user_id)
        
        # Generate comparison data
        comparison_data = await self._generate_comparison_data(comparison.food_ids)
        
        # Create comparison record
        comparison_record = {
            "user_id": user_id,
            "comparison_name": comparison.comparison_name,
            "food_ids": comparison.food_ids,
            "comparison_data": comparison_data
        }
        
        db_service = DatabaseOperationService(self.supabase)
        result = await db_service.insert_with_timestamps("food_comparisons", comparison_record)
        
        return FoodComparisonResponse(**result)
    
    async def get_comparison(
        self, 
        comparison_id: str, 
        user_id: str
    ) -> FoodComparisonResponse:
        """
        Get a food comparison by ID
        
        Args:
            comparison_id: Comparison ID
            user_id: User ID for authorization
            
        Returns:
            Food comparison data
        """
        response = self.supabase.table("food_comparisons")\
            .select("*")\
            .eq("id", comparison_id)\
            .eq("user_id", user_id)\
            .execute()
        
        if not response.data:
            raise ValueError("Comparison not found or access denied")
        
        return FoodComparisonResponse(**response.data[0])
    
    async def get_user_comparisons(
        self, 
        user_id: str,
        limit: int = 20
    ) -> List[FoodComparisonResponse]:
        """
        Get all comparisons for a user
        
        Args:
            user_id: User ID
            limit: Maximum number of comparisons to return
            
        Returns:
            List of user's comparisons
        """
        response = self.supabase.table("food_comparisons")\
            .select("*")\
            .eq("user_id", user_id)\
            .order("created_at", desc=True)\
            .limit(limit)\
            .execute()
        
        return [FoodComparisonResponse(**comp) for comp in response.data]
    
    async def get_comparison_dashboard(
        self, 
        comparison_id: str, 
        user_id: str
    ) -> FoodComparisonDashboard:
        """
        Get comprehensive comparison dashboard data
        
        Args:
            comparison_id: Comparison ID
            user_id: User ID for authorization
            
        Returns:
            Comparison dashboard data
        """
        # Get comparison
        comparison = await self.get_comparison(comparison_id, user_id)
        
        # Get food details
        foods = await self._get_food_details(comparison.food_ids)
        
        # Generate detailed metrics
        metrics = await self._generate_detailed_metrics(comparison.food_ids, foods)
        
        # Generate recommendations
        recommendations = await self._generate_comparison_recommendations(metrics, foods)
        
        # Determine best options
        best_overall = self._determine_best_overall(metrics, foods)
        best_value = self._determine_best_value(metrics, foods)
        best_nutrition = self._determine_best_nutrition(metrics, foods)
        
        return FoodComparisonDashboard(
            comparison_id=comparison_id,
            comparison_name=comparison.comparison_name,
            foods=foods,
            metrics=metrics,
            recommendations=recommendations,
            best_overall=best_overall,
            best_value=best_value,
            best_nutrition=best_nutrition
        )
    
    async def compare_foods_quick(
        self, 
        food_ids: List[str], 
        user_id: str
    ) -> FoodComparisonMetrics:
        """
        Quick comparison of foods without saving
        
        Args:
            food_ids: List of food IDs to compare
            user_id: User ID for authorization
            
        Returns:
            Comparison metrics
        """
        # Validate food access
        await self._validate_food_access(food_ids, user_id)
        
        # Get food details
        foods = await self._get_food_details(food_ids)
        
        # Generate metrics
        return await self._generate_detailed_metrics(food_ids, foods)
    
    async def _validate_food_access(self, food_ids: List[str], user_id: str) -> None:
        """Validate that user has access to all food IDs"""
        for food_id in food_ids:
            response = self.supabase.table("food_analyses")\
                .select("id")\
                .eq("id", food_id)\
                .eq("user_id", user_id)\
                .execute()
            
            if not response.data:
                raise ValueError(f"Food analysis {food_id} not found or access denied")
    
    async def _get_food_details(self, food_ids: List[str]) -> List[Dict[str, Any]]:
        """Get detailed food information for comparison"""
        foods = []
        
        for food_id in food_ids:
            response = self.supabase.table("food_analyses")\
                .select("*")\
                .eq("id", food_id)\
                .execute()
            
            if response.data:
                food_data = response.data[0]
                foods.append({
                    "id": food_data["id"],
                    "name": food_data["food_name"],
                    "brand": food_data.get("brand", "Unknown"),
                    "calories_per_100g": food_data["calories_per_100g"],
                    "protein_percentage": food_data["protein_percentage"],
                    "fat_percentage": food_data["fat_percentage"],
                    "fiber_percentage": food_data["fiber_percentage"],
                    "moisture_percentage": food_data["moisture_percentage"],
                    "ingredients": food_data.get("ingredients", []),
                    "allergens": food_data.get("allergens", []),
                    "analyzed_at": food_data["analyzed_at"]
                })
        
        return foods
    
    async def _generate_comparison_data(self, food_ids: List[str]) -> Dict[str, Any]:
        """Generate comparison data for storage"""
        foods = await self._get_food_details(food_ids)
        metrics = await self._generate_detailed_metrics(food_ids, foods)
        
        return {
            "foods": foods,
            "metrics": metrics.dict(),
            "generated_at": DateTimeService.now_iso()
        }
    
    async def _generate_detailed_metrics(
        self, 
        food_ids: List[str], 
        foods: List[Dict[str, Any]]
    ) -> FoodComparisonMetrics:
        """Generate detailed comparison metrics"""
        if not foods:
            raise ValueError("No food data available for comparison")
        
        # Extract nutritional data
        calories = {food["id"]: food["calories_per_100g"] for food in foods}
        protein = {food["id"]: food["protein_percentage"] for food in foods}
        fat = {food["id"]: food["fat_percentage"] for food in foods}
        fiber = {food["id"]: food["fiber_percentage"] for food in foods}
        
        # Calculate cost per calorie (mock data - would come from price data)
        cost_per_calorie = {food["id"]: 0.01 + (i * 0.005) for i, food in enumerate(foods)}
        
        # Calculate nutritional density (calories per gram of essential nutrients)
        nutritional_density = {}
        for food in foods:
            # Simple nutritional density calculation
            essential_nutrients = food["protein_percentage"] + food["fat_percentage"]
            if essential_nutrients > 0:
                density = food["calories_per_100g"] / essential_nutrients
            else:
                density = 0
            nutritional_density[food["id"]] = round(density, 2)
        
        # Calculate compatibility scores (mock data - would come from pet compatibility)
        compatibility_scores = {food["id"]: 70 + (i * 5) for i, food in enumerate(foods)}
        
        # Generate overall rankings
        overall_rankings = self._calculate_overall_rankings(foods, calories, protein, fat, fiber, cost_per_calorie)
        
        return FoodComparisonMetrics(
            calories_comparison=calories,
            protein_comparison=protein,
            fat_comparison=fat,
            fiber_comparison=fiber,
            cost_per_calorie=cost_per_calorie,
            nutritional_density=nutritional_density,
            compatibility_scores=compatibility_scores,
            overall_rankings=overall_rankings
        )
    
    def _calculate_overall_rankings(
        self, 
        foods: List[Dict[str, Any]], 
        calories: Dict[str, float],
        protein: Dict[str, float],
        fat: Dict[str, float],
        fiber: Dict[str, float],
        cost_per_calorie: Dict[str, float]
    ) -> List[Dict[str, Any]]:
        """Calculate overall rankings for foods"""
        rankings = []
        
        for food in foods:
            food_id = food["id"]
            
            # Calculate composite score (0-100)
            # Weight factors: calories (25%), protein (25%), fat (20%), fiber (15%), cost (15%)
            calorie_score = min(100, max(0, 100 - abs(calories[food_id] - 350) / 3.5))
            protein_score = min(100, protein[food_id] * 2)  # 50% protein = 100 points
            fat_score = min(100, fat[food_id] * 4)  # 25% fat = 100 points
            fiber_score = min(100, fiber[food_id] * 10)  # 10% fiber = 100 points
            cost_score = min(100, max(0, 100 - cost_per_calorie[food_id] * 1000))
            
            composite_score = (
                calorie_score * 0.25 +
                protein_score * 0.25 +
                fat_score * 0.20 +
                fiber_score * 0.15 +
                cost_score * 0.15
            )
            
            rankings.append({
                "food_id": food_id,
                "food_name": food["name"],
                "brand": food["brand"],
                "composite_score": round(composite_score, 1),
                "calorie_score": round(calorie_score, 1),
                "protein_score": round(protein_score, 1),
                "fat_score": round(fat_score, 1),
                "fiber_score": round(fiber_score, 1),
                "cost_score": round(cost_score, 1)
            })
        
        # Sort by composite score
        rankings.sort(key=lambda x: x["composite_score"], reverse=True)
        
        return rankings
    
    async def _generate_comparison_recommendations(
        self, 
        metrics: FoodComparisonMetrics, 
        foods: List[Dict[str, Any]]
    ) -> List[str]:
        """Generate recommendations based on comparison metrics"""
        recommendations = []
        
        if not foods:
            return recommendations
        
        # Find best and worst performers
        best_overall = max(metrics.overall_rankings, key=lambda x: x["composite_score"])
        worst_overall = min(metrics.overall_rankings, key=lambda x: x["composite_score"])
        
        # Protein recommendations
        protein_values = list(metrics.protein_comparison.values())
        if protein_values:
            max_protein = max(protein_values)
            min_protein = min(protein_values)
            
            if max_protein - min_protein > 10:
                recommendations.append(f"Protein content varies significantly ({min_protein:.1f}% - {max_protein:.1f}%)")
        
        # Calorie recommendations
        calorie_values = list(metrics.calories_comparison.values())
        if calorie_values:
            max_calories = max(calorie_values)
            min_calories = min(calorie_values)
            
            if max_calories - min_calories > 100:
                recommendations.append(f"Calorie content varies significantly ({min_calories:.0f} - {max_calories:.0f} kcal/100g)")
        
        # Cost recommendations
        cost_values = list(metrics.cost_per_calorie.values())
        if cost_values:
            max_cost = max(cost_values)
            min_cost = min(cost_values)
            
            if max_cost / min_cost > 2:
                recommendations.append("Significant cost variation - consider value for money")
        
        # Overall recommendation
        if best_overall["composite_score"] > 80:
            recommendations.append(f"{best_overall['food_name']} scores highest overall ({best_overall['composite_score']:.1f}/100)")
        elif worst_overall["composite_score"] < 40:
            recommendations.append(f"Consider avoiding {worst_overall['food_name']} (score: {worst_overall['composite_score']:.1f}/100)")
        
        return recommendations
    
    def _determine_best_overall(self, metrics: FoodComparisonMetrics, foods: List[Dict[str, Any]]) -> str:
        """Determine best overall food"""
        if not metrics.overall_rankings:
            return "No data"
        
        best = metrics.overall_rankings[0]
        return f"{best['food_name']} ({best['composite_score']:.1f}/100)"
    
    def _determine_best_value(self, metrics: FoodComparisonMetrics, foods: List[Dict[str, Any]]) -> str:
        """Determine best value food"""
        if not metrics.cost_per_calorie:
            return "No data"
        
        # Find food with lowest cost per calorie
        best_value_id = min(metrics.cost_per_calorie.keys(), key=lambda x: metrics.cost_per_calorie[x])
        best_food = next((f for f in foods if f["id"] == best_value_id), None)
        
        if best_food:
            cost = metrics.cost_per_calorie[best_value_id]
            return f"{best_food['name']} (${cost:.3f}/kcal)"
        
        return "No data"
    
    def _determine_best_nutrition(self, metrics: FoodComparisonMetrics, foods: List[Dict[str, Any]]) -> str:
        """Determine best nutritional food"""
        if not metrics.overall_rankings:
            return "No data"
        
        # Find food with highest nutritional density
        best_nutrition_id = max(metrics.nutritional_density.keys(), key=lambda x: metrics.nutritional_density[x])
        best_food = next((f for f in foods if f["id"] == best_nutrition_id), None)
        
        if best_food:
            density = metrics.nutritional_density[best_food["id"]]
            return f"{best_food['name']} (density: {density:.1f})"
        
        return "No data"
    
    async def delete_comparison(self, comparison_id: str, user_id: str) -> bool:
        """Delete a food comparison"""
        response = self.supabase.table("food_comparisons")\
            .delete()\
            .eq("id", comparison_id)\
            .eq("user_id", user_id)\
            .execute()
        
        return len(response.data) > 0
